library/java/org/chromium/net/impl/CronetBidirectionalStream.java

```diff
@@ -233,163 +176,160 @@ public void run() {
+      if (!mDelayRequestHeadersUntilFirstFlush) {
+        startingEvent |= Event.USER_START_WITH_HEADERS;
+      }
+      if (mReadOnly) {
+        startingEvent |= Event.USER_START_READ_ONLY;
       }
-      try {
-        mNativeStream = CronetBidirectionalStreamJni.get().createBidirectionalStream(
-            CronetBidirectionalStream.this, mRequestContext.getEnvoyEngine(),
-            !mDelayRequestHeadersUntilFirstFlush, mRequestContext.hasRequestFinishedListener(),
-            mTrafficStatsTagSet, mTrafficStatsTag, mTrafficStatsUidSet, mTrafficStatsUid);
-        mRequestContext.onRequestStarted();
-        // Non-zero startResult means an argument error.
-        int startResult = CronetBidirectionalStreamJni.get().start(
-            mNativeStream, CronetBidirectionalStream.this, mInitialUrl, mInitialPriority,
-            mInitialMethod, mRequestHeaders, !doesMethodAllowWriteData(mInitialMethod));
-        if (startResult == -1) {
-          throw new IllegalArgumentException("Invalid http method " + mInitialMethod);
-        }
-        if (startResult > 0) {
-          int headerPos = startResult - 1;
-          throw new IllegalArgumentException("Invalid header " + mRequestHeaders[headerPos] + "=" +
-                                             mRequestHeaders[headerPos + 1]);
+    } else {
+      startingEvent = Event.ERROR;
+    }
+    @NextAction int nextAction = mState.nextAction(startingEvent);
+    mStartMillis = SystemClock.elapsedRealtime();
+    mRequestContext.onRequestStarted();
+    if (nextAction == NextAction.PROCESS_ERROR) {
+      mException.set(startUpException);
+      failWithException();
+      return;
+    }
+    try {
+      mRequestContext.setTaskToExecuteWhenInitializationIsCompleted(new Runnable() {
+        @Override
+        public void run() {
+          mStartBlock.open();
         }
-        mReadState = mWriteState = State.STARTED;
-      } catch (RuntimeException e) {
-        // If there's an exception, clean up and then throw the
-        // exception to the caller.
-        destroyNativeStreamLocked(false);
-        throw e;
-      }
+      });
+      mStartBlock.block();
+      mStream.setStream(
+          mRequestContext.getEnvoyEngine().startStream(this, /* explicitFlowCrontrol= */ true));
+      if (nextAction == NextAction.FLUSH_HEADERS) {
+        mStream.sendHeaders(mEnvoyRequestHeaders, mReadOnly);
+      }
+      onStreamReady();
+    } catch (RuntimeException e) {
+      // Will be reported when "onCancel" gets invoked.
+      reportException(new CronetExceptionImpl("Startup failure", e));
+    }
+  }
+
+  /**
+   * Returns, potentially, an exception to report through the "onError" callback, even though no
+   * stream has been created yet. This awkward error reporting solely exists to mimic Cronet.
+   */
+  @Nullable
+  private static CronetException engineSimulatedError(Map<String, List<String>> requestHeaders) {
+    if (requestHeaders.get(":scheme").get(0).equals("http")) {
```

Comment:
That's implemented in C++, I have not yet paid serious attention to this. There is a test that trickles this error. As usual, it is not because a test is passing that we can imply that this is the correct implementation. Far from it. I will add to TODO.